apps/src/lab2/projects/ProjectManager.ts

```diff
@@ -357,12 +357,13 @@ export default class ProjectManager {
   private onSaveFail(errorMessage: string, error: Error) {
     this.saveInProgress = false;
     this.executeSaveFailListeners(error);
-    if (error.message.includes('409')) {
-      // If this is a conflict, we need to reload the page.
+    if (error.message.includes('409') || error.message.includes('401')) {
```

Comment:
(this is a future task) I wonder if instead of force reloading, especially in the unauthorized case, we should do some sort of pop-up telling the user we can't save so they can save/look at their code before it disappears? We had issues with lost progress this week because the browser forced a reload. I don't know if there's a way for us to recover without reloading, but we could at least try to help the user save their progress elsewhere before we reload?

Reply:
Yeah good call, that would be ideal. Maybe a way for them to one-click remix or copy their code so they can paste into a new project? That being said I think in theory the worst case impact here is that the user would lose their last 30s of work, because that's the longest they could go after a code change was made without saving due to autosave? But yes agreed that we should handle this in a more elegant way.